dags/nyc-taxi-dag.py

```diff
@@ -0,0 +1,138 @@
+import datetime
+import logging
+
+from pathlib import Path
+from airflow import DAG
+from airflow.providers.postgres.operators.postgres import PostgresOperator
+from airflow.hooks.postgres_hook import PostgresHook
+from airflow.operators.http_operator import SimpleHttpOperator
+from airflow.operators.python_operator import PythonOperator
+
+
+def get_processed_files(ti):
+    pg_hook = PostgresHook(postgres_conn_id="cratedb_demo_connection")
+    records = pg_hook.get_records(sql='SELECT file_name FROM nyc_taxi.load_files_processed')
+
+    # flatten nested list as there is only one column
+    return list(map(lambda record: record[0], records))
+
+
+def clean_data_urls(ti):
+    data_urls_raw = ti.xcom_pull(task_ids="download_data_urls", key="return_value")
+
+    data_urls = data_urls_raw.split('\n')
+    # we only import Yellow tripdata for now due to different CSV schemas
+    data_urls_filtered = filter(lambda element: 'yellow' in element, data_urls)
+
+    return list(data_urls_filtered)
+
+
+def identitfy_missing_urls(ti):
+    data_urls_processed = ti.xcom_pull(task_ids="get_processed_files")
+    data_urls_available = ti.xcom_pull(task_ids="clean_data_urls")
+
+    return list(set(data_urls_available) - set(data_urls_processed))
+
+
+def process_new_files(ti):
+    missing_urls = ti.xcom_pull(task_ids="identitfy_missing_urls")
+
+    for missing_url in missing_urls:
+        logging.info(missing_url)
+
+        file_name = missing_url.split('/').pop()
+
+        PostgresOperator(
+            dag=dag,
+            task_id="copy_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql="""
+                    COPY nyc_taxi.load_trips_staging
+                    FROM '{url}'
+                    WITH (format = 'csv', empty_string_as_null = true)
+                    RETURN SUMMARY;
+                """.format(
+                url=missing_url
+                )
+        ).execute(dict())
+
+        PostgresOperator(
+            dag=dag,
+            task_id="log_{url}".format(url=file_name),
+            postgres_conn_id="cratedb_demo_connection",
+            sql=Path('/usr/local/airflow/include/taxi-insert.sql').read_text(),
```

Comment:
This is not the preferred solution (the operator should deal with reading the file and just receive the file name). But there appears to be a limitation if a `PostgresOperator` is called from within a `PythonOperator`: https://stackoverflow.com/questions/67229594/airflow-running-postgresoperator-within-a-pythonoperator-with-taskflow-api

Reply:
I am wondering is there a way to specify a relative path?

Reply:
Yes, `Path('include/taxi-insert.sql')` also worked, changed that. Initially I thought the absolute path shouldn't hurt as we are inside the Docker container and there the path should be identical for all setups. But a relative path is better anyway.